base/lru_cache.hpp

```diff
@@ -0,0 +1,164 @@
+  /// For the same |key| should be loaded the same |value|.
+  LruCache(size_t maxCacheSize, Loader const & loader)
+    : m_maxCacheSize(maxCacheSize), m_loader(loader)
+  {
+    CHECK_GREATER(maxCacheSize, 0, ());
+  }
+
+  /// \brief Loads value, if it's necessary, by |key| with |m_loader|, puts it to |m_cache| and
+  /// returns the reference to the value to |m_cache|.
+  Value const & GetValue(Key const & key)
+  {
+    auto const it = m_cache.find(key);
+    if (it != m_cache.cend())
+    {
+      m_keyAge.UpdateAge(key);
+      return it->second;
+    }
+
+    if (m_cache.size() >= m_maxCacheSize)
+    {
+      m_cache.erase(m_keyAge.GetLruKey());
+      m_keyAge.RemoveLru();
+    }
+
+    m_keyAge.InsertKey(key);
+    Value & value = m_cache[key];
+    m_loader(key, value);
+    return value;
+  }
+
+  /// \brief Checks for coherence class params.
+  /// \note It's a time consumption method and should be called for tests only.
+  bool IsValid() const
+  {
+    if (!m_keyAge.IsValidForTesting())
+      return false;
+
+    for (auto const & kv : m_cache)
+    {
+      if (!m_keyAge.IsKeyValidForTesting(kv.first))
+        return false;
+    }
+
+    return true;
+  }
+
+private:
+  /// \brief This class support cross mapping from age to key and for key to age.
+  /// It lets effectively get least recently used key (key with minimum value of age)
+  /// and find key age by its value to update the key age.
+  /// \note Size of |m_ageToKey| and |m_ageToKey| should be the sameß.
+  /// All keys of |m_ageToKey| should be values of |m_ageToKey| and on the contrary
+  /// all keys of |m_ageToKey| should be values of |m_ageToKey|.
+  /// \note Ages should be unique for all keys.
+  class KeyAge
+  {
+    template <typename K, typename V> friend class LruCacheKeyAgeTest;
+  public:
+    /// \brief Increments |m_age| and insert key to |m_ageToKey| and |m_keyToAge|.
+    /// \note This method should be used only if there's no |key| in |m_ageToKey| and |m_keyToAge|.
+    void InsertKey(Key const & key)
+    {
+      ++m_age;
+      m_ageToKey[m_age] = key;
+      m_keyToAge[key] = m_age;
+    }
+
+    /// \brief Increments |m_age| and updates key age in |m_ageToKey| and |m_keyToAge|.
+    /// \note This method should be used only if there's |key| in |m_ageToKey| and |m_keyToAge|.
+    void UpdateAge(Key const & key)
+    {
+      ++m_age;
+      auto keyToAgeIt = m_keyToAge.find(key);
+      CHECK(keyToAgeIt != m_keyToAge.end(), ());
+      // Removing former age.
+      size_t const removed = m_ageToKey.erase(keyToAgeIt->second);
+      CHECK_EQUAL(removed, 1, ());
+      // Putting new age.
+      m_ageToKey[m_age] = key;
+      keyToAgeIt->second = m_age;
+    }
+
+    /// \returns Least recently used key without updating the age.
+    /// \note |m_ageToKey| and |m_keyToAge| shouldn't be empty.
+    Key const & GetLruKey() const
+    {
+      CHECK(!m_ageToKey.empty(), ());
```

Comment:
Not sure, I don't remember where I implemented LRU - in maps.me repository or in my own library, but you can easily implement LRU cache with expected O(1) for insert, remove, update. All what you need is following data structures:
* list of keys, stored in the LRU order
* hash map from key to the tuple (value, iterator to the same key in the list)

So, when you insert new key, all what you need is to push new item to the beginning of the list and insert a new entry in the hash map..

When you update existing key, you need to just move corresponding list item to the beginning, and update iterators.

When you need to evict an item from the cache, just remove the last item of the list and remove corresponding entry from the hash-map.

Reply:
As I said, probably this implementation already exists somewhere in the search/ directory, but I'm not sure, need to check this.

Reply:
Thanks a lot. I'll try to implement that measure again.

Reply:
This was a good idea...